Make RenderSVGResourceClipper lazily recompute clip-path data when referencing element is changed

heycam opened the pull request:
#### ce878379210ced398d994217e26519b433761149
<pre>
Make RenderSVGResourceClipper lazily recompute clip-path data when referencing element is changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=242423">https://bugs.webkit.org/show_bug.cgi?id=242423</a>
&lt;rdar://problem/96571459&gt;

Reviewed by Said Abou-Hallawa.

Like <a href="https://bugs.webkit.org/show_bug.cgi?id=242421">https://bugs.webkit.org/show_bug.cgi?id=242421</a> did for SVG gradient
resources, this patch gives RenderSVGResourceClipper the ability to
lazily invalidate its cached data for a given client by storing its
dependencies and validating them in applyClippingToContext.

* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
(WebCore::RenderSVGResourceClipper::removeAllClientsFromCache):
(WebCore::RenderSVGResourceClipper::removeClientFromCache):
(WebCore::RenderSVGResourceClipper::computeInputs):
(WebCore::RenderSVGResourceClipper::applyClippingToContext):
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
(WebCore::RenderSVGResourceClipper::addRendererToClipper): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.h:
(WebCore::ClipperData::Inputs::operator==):
(WebCore::ClipperData::validate):
* Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
(WebCore::hasResourcesRequiringRemovalOnClientLayoutChange):

Canonical link: <a href="https://commits.webkit.org/252311@main">https://commits.webkit.org/252311@main</a>
</pre>

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -160,14 +151,28 @@ bool RenderSVGResourceClipper::applyClippingToContext(GraphicsContext& context,
+    if (!clipperData.imageBuffer && pathOnlyClipping(context, animatedLocalTransform, objectBoundingBox, effectiveZoom))
+        return true;
+
+    if (clipperData.invalidate(computeInputs(renderer, objectBoundingBox, clippedContentBounds, effectiveZoom))) {
```
I am confused here. We used to regenerate the maskImage if absoluteTransform changes. Now we look at the changes in effectiveZoom. effectiveZoom does not affect the size of the maskImage but absoluteTransform can affect the size of the maskImage since we get the scale out of it

```
    // Ignore 2D rotation, as it doesn't affect the size of the mask.
    FloatSize scale(absoluteTransform.xScale(), absoluteTransform.yScale());
```

heycam replied:
We store both the `effectiveZoom` and the scale extracted from the `absoluteTransform` (and clamped by the objectBoundingBox size) in the `Inputs`, so we invalidate the mask image if either of those change.

shallawa replied:
Okay sorry. I was confused by absoluteTransform and animatedLocalTransform.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.h:
```diff
@@ -30,6 +30,32 @@ class GraphicsContext;
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+
+    struct Inputs {
+        bool operator==(const Inputs& other) const { return objectBoundingBox == other.objectBoundingBox && clippedContentBounds == other.clippedContentBounds && scale == other.scale && effectiveZoom == other.effectiveZoom; }
```
Maybe this can be more readable:

```
bool operator==(const Inputs& other) const
{
    return std::tie(objectBoundingBox, clippedContentBounds, scale, effectiveZoom) == std::tie(other.objectBoundingBox, other.clippedContentBounds, other.scale, other.effectiveZoom);
}
```

heycam replied:
Nice trick!

heycam committed:
Make RenderSVGResourceClipper lazily recompute clip-path data when referencing element is changed
https://bugs.webkit.org/show_bug.cgi?id=242423
<rdar://problem/96571459>

Reviewed by Said Abou-Hallawa.

Like https://bugs.webkit.org/show_bug.cgi?id=242421 did for SVG gradient
resources, this patch gives RenderSVGResourceClipper the ability to
lazily invalidate its cached data for a given client by storing its
dependencies and validating them in applyClippingToContext.

* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
(WebCore::RenderSVGResourceClipper::removeAllClientsFromCache):
(WebCore::RenderSVGResourceClipper::removeClientFromCache):
(WebCore::RenderSVGResourceClipper::computeInputs):
(WebCore::RenderSVGResourceClipper::applyClippingToContext):
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
(WebCore::RenderSVGResourceClipper::addRendererToClipper): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.h:
(WebCore::ClipperData::Inputs::operator==):
(WebCore::ClipperData::validate):
* Source/WebCore/rendering/svg/SVGResourcesCache.cpp:
(WebCore::hasResourcesRequiringRemovalOnClientLayoutChange):

Canonical link: https://commits.webkit.org/252311@main
